src/target/riscv/riscv.c

```diff
@@ -2278,6 +2279,67 @@ derive_debug_reason_without_hitbit(const struct target *target, riscv_reg_t dpc)
 	}
 	return DBG_REASON_WATCHPOINT;
 }
+
+static bool check_single_step_needed_in_hit_supported_case(struct target *target,
```

Comment:
Please, consider moving it back to the previous location inside `hit` detection. This will make the change/restore of `tselect` and the read of `tdata1` unnecessary.

Reply:
addressed